File: LeetCode/ProblemSet/Volume (1300-1399)/LeetCode1331/LeetCode1331-Solution-01.go
package leetcode1331

import (
	"sort"
)

func arrayRankTransform(arr []int) []int {
	a := append([]int{}, arr...)
	sort.Ints(a)

	book := make(map[int]int)
	for _, v := range a {
		if _, exist := book[v]; !exist {
			book[v] = len(book) + 1
		}
	}

	ans := make([]int, len(arr))
	for i, v := range arr {
		ans[i] = book[v]
	}

	return ans
}

File: LeetCode/ProblemSet/Volume (1300-1399)/LeetCode1331/LeetCode1331-Solution-01.cpp
#include <bits/stdc++.h>

class Solution {
public:
    std::vector<int> arrayRankTransform(std::vector<int>& arr) {
        std::vector<int> a(arr);
        std::ranges::sort(a);
        
        std::unordered_map<int, int> book;
        for (int v : a) {
            if (!book.contains(v)) {
                book[v] = book.size() + 1;
            }
        }

        std::vector<int> ans;
        for (int v : arr) {
            ans.emplace_back(book[v]);
        }

        return ans;
    }
};